Bottom: 95b738709159d4b1b7e2425ff9ce15d2e4edf930
Top:    c2f2d432b6119b89dd478cc0bafb0313c6cdd5d2
Author: = <=>
Date:   2016-12-07 18:04:46 +0800

macvtap: submit skb list for host kernel stack

Signed-off-by: = <=>


---

diff --git a/drivers/net/macvtap.c b/drivers/net/macvtap.c
index 2235d9d..d3bf19e 100644
--- a/drivers/net/macvtap.c
+++ b/drivers/net/macvtap.c
@@ -686,8 +686,18 @@ static int macvtap_batch_xmit(struct macvtap_queue *q, struct sk_buff *skb,
 	spin_unlock(&queue->lock);
 
 	if (rcv) {
-		while ((skb = __skb_dequeue(&process_queue)))
-			dev_queue_xmit(skb);
+		struct sk_buff *tmp = NULL;
+		struct sk_buff *head = NULL;
+		while ((skb = __skb_dequeue(&process_queue))) {
+			if (!tmp)
+				tmp = head = skb;
+			else {
+				tmp->next = skb;
+				tmp = skb;
+			}
+		}
+		tmp->next = NULL;
+		dev_queue_xmit(head);
 	}
 
 	return 0;
